examples/speculative/speculative.cpp

```diff
@@ -177,6 +178,27 @@ int main(int argc, char ** argv) {
             llama_eval(ctx_dft, &id, 1, n_past_dft, params.n_threads);
             ++n_past_dft;
 
+            // heuristic for n_draft
+            {
+                const int  n_draft_cur  = (int) drafted.size();
+                const bool all_accepted = i_dft == n_draft_cur;
+
+                LOG("n_draft      = %d\n", n_draft);
+                LOG("n_draft_cur  = %d\n", n_draft_cur);
+                LOG("i_dft        = %d\n", i_dft);
+                LOG("all_accepted = %d\n", all_accepted);
+
+                if (all_accepted && n_draft == n_draft_cur) {
+                    LOG(" - max drafted tokens accepted - n_draft += 8\n");
+                    n_draft = std::min(30, n_draft + 8);
+                } else if (all_accepted) {
+                    LOG(" - partially drafted tokens accepted - no change\n");
+                } else {
+                    LOG(" - drafted token rejected - n_draft -= 1\n");
+                    n_draft = std::max(2, n_draft - 1);
+                }
+            }
+
```

Comment:
@leng-yue 

I've refactored the implementation to be more contained.
Also, we were rewarding the draft even when it hasn't sampled all `n_draft` tokens, which does not seem correct.

For example, let's say `n_draft` is 16 currently, but the draft samples just 3 tokens because the "low-probability" check has been triggered. Even if all 3 tokens were accepted, we should not reward the draft model, because this is just a small part of what we asked it to do.

Regarding the reproducibility - we will study this more in #3014 
My guess is that this behavior would occur even without the heuristic - probably it's just less likely to happen for some reason when the heuristic is disabled.

Reply:
In my earlier implementation, rewards were given only when all tokens were accepted. So if only 3 out of 16 tokens are accepted, the n_tokens value would be decreased by 1.

Reply:
Not exactly. This check does not guarantee you have `n_draft` tokens accepted:

https://github.com/ggerganov/llama.cpp/blob/98230ef656641597007d46854e005d16ff677377/examples/speculative/speculative.cpp#L146-L148

The reason is because `drafted.size() <= n_draft` due to another heuristic of not drafting more tokens if the drafter becomes "unsure":

https://github.com/ggerganov/llama.cpp/blob/98230ef656641597007d46854e005d16ff677377/examples/speculative/speculative.cpp#L192-L196

So in the majority of cases, when `all_accepted == true`, you would have accepted less than `n_draft` tokens. That's why your `n_draft` would increase so much even beyond 70 in some cases.

Reply:
Got it.